File: PA1 - Integer List ADT/src/Lex.java
/**
 * Davie Truong
 * dtruong8@ucsc.edu
 * 1/15/2017
 * CMPS 101 Algorithms and Abstract Data Types - Patrick Tantalo
 * Programming Assignment 1 - Lex Implementation file
 */

import java.io.*;


public class Lex {

    public static void main(String[] args){
        if(args.length != 2){       // check to see if the user inputted two command line args
            System.err.println("You must open lex.java with two file arguments, input.txt and output.txt");
            System.exit(0);
        }

        String inputFile = args[0];
        String outputFile = args[1];
        int linesInFile = 0;
        String line;

        // Opening the file to get the number of lines in the file
        try{
            BufferedReader input = new BufferedReader(new FileReader(inputFile));
            while((input.readLine()) != null){
                linesInFile += 1;
            }
            input.close();
        }catch(Exception e){
            System.out.println("Input File Not Found");
        }

        // Make a String array with size number of words in the file
        String [] wordArr = new String [linesInFile];
        // Opening the file to put the words into the array
        try{
            BufferedReader input = new BufferedReader(new FileReader(inputFile));
            int i = 0;
            while((line = input.readLine()) != null){
                wordArr[i] = line;
                i++;
            }
            input.close();
        }catch(Exception e){
            System.out.println("Input File Not Found");
        }
        // Gets the array of words and uses the list ADT to sort the array of words in alphabetical order without changing it
        List intList = new List();  // creates an empty list
        intList.append(0);          // append the first word/int to the list so there is a starting point
        for(int i = 1; i < linesInFile; i++){
            intList.moveFront();    // sets the cursor at the front of the list\
            boolean completeInsertion = false;  // used as a loop control to determine if the index gets added
            while(completeInsertion == false)  // if the insertion has not been made
            {
                if(intList.index() < 0){        // checks to see if the cursor fell of the list
                    intList.append(i);     // if the cursor fell of the list it means the data is greatest value of list
                    completeInsertion = true;   // set the loop control to true so it can exit out and start next data
                }else if(wordArr[intList.get()].compareTo(wordArr[i]) > 0){   // checks to see if data is less than current
                intList.insertBefore(i);        // insert before because the data is less then current index data
                completeInsertion = true;       // set loop control to true so it can exit and start on next data
                }else{
                intList.moveNext();   // moves the index next because the data is greater so can't insert before
                }
            }
        }
        // Opens the output file to store the user's words in alphabetical order
        try{
            PrintWriter output = new PrintWriter(outputFile);
            intList.moveFront();
            for(int i = 0; i < linesInFile; i++){
                output.println(wordArr[intList.get()]);
                intList.moveNext();
            }
            output.close();
        }catch(Exception e){
            System.out.println("Output File Not Found");
        }
    }
}

File: PA2 - Integer List ADT C/PA2 - Integer List ADT C/Lex.c
/**
* Davie Truong
* dtruong8@ucsc.edu
* 1/22/2017
* CMPS 101 Algorithms and Abstract Data Types - Patrick Tantalo
* Programming Assignment 2 - Lex Implementation file in C
*/

#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include"List.h"
#define MAX_LEN 160


int main(int argc, char* argv[]) {

	int lineCount = 0;
	FILE *in, *out;
	char wordString[MAX_LEN];

	//// check command line for correct number of arguments
	if (argc != 3) {
		printf("Usage: %s <input file> <output file>\n", argv[0]);
		exit(1);
	}


	// open files for reading and writing 
	in = fopen(argv[1], "r");
	out = fopen(argv[2], "w");
	if (in == NULL) {
		printf("Unable to open file %s for reading\n", argv[1]);
		exit(1);
	}
	if (out == NULL) {
		printf("Unable to open file %s for writing\n", argv[2]);
		exit(1);
	}

	// Data Gathering Part of Lex
	// read each line of input file to get the count
	while (fgets(wordString, MAX_LEN, in) != NULL) {
		lineCount++;
	}

	//char wordArray[lineCount][MAX_LEN];	// makes the array to store the word
	char wordArray[20][MAX_LEN];	// makes the array to store the word
	rewind(in);							// start at the beginning of the input file
	int i = 0;							// used to iterate the array
	// this loop stores the words from the file into the array 
	while (fgets(wordString, MAX_LEN, in) != NULL) {
		strcpy(wordArray[i], wordString);
		i++;
	}
	fclose(in);							// close the input file

	// Sorting Part of Lex
	List intList = newList();			// Creates an empty List
	append(intList, 0);					// append the first word/int to the list so there is a starting point
	for (i = 1; i < lineCount; i++) {
		moveFront(intList);				// sets the cursor at the front of the list
		int completeInsertion = 0;			// used as a loop control to determne if the index gets added
		while (completeInsertion == 0) {		// if the insertion has not been made
			if (index(intList) < 0) {			// checks to see if the cursor fell of the list
				append(intList, i);				// if the cursor fell of the list it means the data is greatest value of list 
				completeInsertion = 1;			// set loop control to true so it can exit out and start next data
			}
			else if (strcmp(wordArray[get(intList)], wordArray[i]) > 0) {	// checks to see if the data is less than current
				insertBefore(intList, i);	// insert before because the data is less than current index data
				completeInsertion = 1;		// set loop control to true so it can exit and start on next data
			}
			else {
				moveNext(intList);	// moves the index next because the data is greater so can't insert before
			}
		}
	}

	// Storing the sorted words into an output list
	moveFront(intList);
	for (i = 0; i < lineCount; i++) {
		fprintf(out, "%s" ,wordArray[get(intList)]);
		moveNext(intList);
	}
	freeList(&intList);	// frees the list from memory
	fclose(out);		// closes the output file

	return (0);
}